Makefile: Remove undefined variable reference

pamolloy opened the pull request.

pamolloy commented:
CC @gmalysa

gmalysa commented:
I think you should submit this one upstream, either add my SoB or I can give you a reviewed tag on the mailing list. It probably needs a little bit more in the commit message itself though, to explain that there is no separate env provided by the LDR files anymore and we use either the default env or the env stored on the device in flash/emmc as appropriate.

pamolloy commented:
> I think you should submit this one upstream, either add my SoB or I can give you a reviewed tag on the mailing list. It probably needs a little bit more in the commit message itself though, to explain that there is no separate env provided by the LDR files anymore and we use either the default env or the env stored on the device in flash/emmc as appropriate.

Absolutely, the goal of all the PRs currently in this repo is to get a review here, then submit mainline, once merged in mainline then `git cherry-pick -x` from mainline and finally merge the PR. Then with every new upstream release we'll create a new branch, rebase/cherry-pick the commits, which should get dropped automatically. Of course there will likely have to be exceptions to that where we merge into `adi-u-boot-*` without changes being mainlined.

Thanks for the context about the environment. It was difficult to dig up any history. Is there any public documentation explaining the environment? Or even internal? We just had a meeting about the ldr utility, which we'll hopefully clean up a bit in the coming months.

gmalysa commented:
> Thanks for the context about the environment. It was difficult to dig up any history. Is there any public documentation explaining the environment? Or even internal? We just had a meeting about the ldr utility, which we'll hopefully clean up a bit in the coming months.

Good to see that ldr will be cleaned up. It seems a bit big (source wise) for such a straightforward file format, but neither nathan nor I wanted to unravel it at the time and just used it as is. Some of the feedback from Tom when we submitted docker changes to include ldr indicated he might want to see support for the format in binman rather than as an external utility, but my opinion is still that yocto should be the meta build tool and should include more thorough packaging rather than relying on uboot to perform packaging of files that are built in different places. If it were reduced to a python lib using pyelftools, then we could easily have it as a standalone tool and trivially integrate it with binman without needing to duplicate anything.

Re: environment, the goal is to move towards using uboot standard boot. We already have all the env variables since they're basically the same as what most legacy boots used, we just haven't committed to the kconfig and remaining dts config yet (which will also improve the secured version since the env isn't signed but the dtb would be part of a signed fit or signed uboot image), and things are done with individual bootcmds, e.g. spiboot, nfsboot, mmcboot. Historically the env was defined in the include/configs/<sc5xx platform> headers, but when we cleaned that up to the very limited form that it is now, we also migrated to the text env which was recommended for mainline. The documentation is pretty minimal but the only instructions are at the top of include/env/adi/adi_boot.h where it says which env variables a board needs to add in order to boot. It runs through the C preprocessor as well, so the board .env file should #define whatever options it wants to enable, like USE_MMC, or USE_SPI.

pamolloy commented:
> Good to see that ldr will be cleaned up. It seems a bit big (source wise) for such a straightforward file format, but neither nathan nor I wanted to unravel it at the time and just used it as is. Some of the feedback from Tom when we submitted docker changes to include ldr indicated he might want to see support for the format in binman rather than as an external utility, but my opinion is still that yocto should be the meta build tool and should include more thorough packaging rather than relying on uboot to perform packaging of files that are built in different places. If it were reduced to a python lib using pyelftools, then we could easily have it as a standalone tool and trivially integrate it with binman without needing to duplicate anything.

Unfortunately, there aren't resources for a rewrite, but it would at least be good, as you suggest, to remove unused files, and then also add CI/CD, code linting, Github releases, rename the repo, update the README, etc.

> Re: environment, the goal is to move towards using uboot standard boot. We already have all the env variables since they're basically the same as what most legacy boots used, we just haven't committed to the kconfig and remaining dts config yet (which will also improve the secured version since the env isn't signed but the dtb would be part of a signed fit or signed uboot image), and things are done with individual bootcmds, e.g. spiboot, nfsboot, mmcboot. Historically the env was defined in the include/configs/ headers, but when we cleaned that up to the very limited form that it is now, we also migrated to the text env which was recommended for mainline. The documentation is pretty minimal but the only instructions are at the top of include/env/adi/adi_boot.h where it says which env variables a board needs to add in order to boot. It runs through the C preprocessor as well, so the board .env file should #define whatever options it wants to enable, like USE_MMC, or USE_SPI.

Ahhh, I didn't mean the U-Boot environment in general, but what the environment referred to by `CREATE_LDR_ENV` was. At first glance it reads like a special environment for LDR, perhaps to configure it or create a build environment. "Create LDR environment" as opposed to "Add U-Boot environment to LDR file". But looking at the removed code makes it clear:

```
-CREATE_LDR_ENV = tools/envcrc --binary > env-ldr.o
```

So the above created a U-Boot environment as an object file so that it could be included in a LDR file with the following flag to LDR:

```
-LDR_FLAGS-$(CONFIG_ENV_IS_EMBEDDED_IN_LDR) += \
-       --punchit $$(($(CONFIG_ENV_OFFSET))):$$(($(CONFIG_ENV_SIZE))):env-ldr.o
-endif
```

And that was all done before realizing that the U-Boot environment didn't need to be stored in the LDR image since there is a static copy in U-Boot proper or optionally an environment on storage somewhere (as opposed to `IS_NOWHERE`). Now I think I understand the history well enough to add some context to the commit message. :smile:

pamolloy committed:
Makefile: Remove expansion of undefined variable

The U-Boot environment was previously stored in the boot loader stream
using ldr, but it has been replaced by the default environment built
into the U-Boot executable or an environment on external storage

Fixes: ea3310e8aafa ("Blackfin: Remove")

Signed-off-by: Philip Molloy <philip.molloy@analog.com>
Reviewed-by: Greg Malysa <malysagreg@gmail.com>
(cherry picked from commit 007610da2cca405ea7739fc120d90085be4b6ac2)

pamolloy commented:
https://lore.kernel.org/u-boot/20250728150333.1716886-1-philip@philipmolloy.com/T/#t

gmalysa commented:
> Ahhh, I didn't mean the U-Boot environment in general, but what the environment referred to by CREATE_LDR_ENV was. At first glance it reads like a special environment for LDR, perhaps to configure it or create a build environment. "Create LDR environment" as opposed to "Add U-Boot environment to LDR file".

Ah yes. I think it was from buildroot or possibly a pre-timesys yocto preparation, where the environment contents would be created from build metadata that was perhaps not known to the uboot include/configs headers. But both uboot and the build system have evolved tremendously since then (circa 2015) so we were able to just make it work normally over time instead.

The result there is similar to how we handled packaging optee.bin for the LDR files, but I put in more awkward use of binutils to get a "proper" ELF file with correct load addresses so that --punchit is not required.

pamolloy commented:
```
Applied to u-boot/master, thanks!

[1/1] Makefile: Remove expansion of undefined variable
      commit: 007610da2cca405ea7739fc120d90085be4b6ac2
```

https://source.denx.de/u-boot/u-boot/-/commit/007610da2cca405ea7739fc120d90085be4b6ac2
